### PR TITLE
add state change tracking

### PR DESCRIPTION
**Pull request by Kelerchian**
This guarantees `.next()` (and consequently for await) will resolve with `done: false, value: not null` ONLY when:
- Caught up for the first time
- Caught up after a state change
- Caught up after a time travel
- Command publishing fails

### CODE REVIEW

**Review by rkuhn — COMMENTED** [review]
This goes into the right direction. Not sure yet about the rules:
- why should we emit the same state again? (e.g. when event processing doesn’t lead to change)
- why should we emit when a command failed? (user should `await` the promise, which will throw an exception)

Wouldn’t it be easier to understand if it only emits when there is a new state, i.e. different from previous?

File: machine-runner/src/runner/runner.ts [hunk 1]
```diff
@@ -336,6 +337,7 @@ export const createMachineRunnerInternal = <
               // Effects of handlingReport on emitters
               ;(() => {
                 if (pushEventResult.executionHappened) {
+                  internals.stateChanged = true
```
[the remark this answers is not in the document]
Comment (Kelerchian) [reply machine-runner/src/runner/runner.ts:340]: If that's the case, the consequence will be that `reactIntoSelf` not trigger the iteration. Then code like this will have to change.

```
if (state.is(Bidding) && state.bids.length > 3) {
   // do stuff only if there are more than 3 bids
}
```

File: machine-runner/src/runner/runner-utils.ts [hunk 2]
```diff
@@ -37,6 +37,7 @@ export type MachineEmitterEventMap<SwarmProtocolName extends string, MachineName
     nextState: unknown
   }) => unknown
   change: (_: StateOpaque<SwarmProtocolName, MachineName>) => unknown
+  caughtUpAndChanged: (_: StateOpaque<SwarmProtocolName, MachineName>) => unknown
```
[the remark this answers is not in the document]
Comment (Kelerchian) [reply machine-runner/src/runner/runner-utils.ts:40]: `change` which I assume is used when machine runner is used as an observable, retains its old behaviors basically so that it is usable in a reactive scenario e.g.:
- after a command is issued, but before the events are published, "change" is emitted. `machine.get()` will then return a nullified `state.commands`.
- on subscription error. `machine.get()` would then return null again
- on every caught up.

**Comment by Kelerchian** [comment]
> why should we emit when a command failed? (user should await the promise, which will throw an exception)

My first thought was because it is neat if the circuit will retry automatically when a command fails, but then on hindsight, if there's a side-effect, the side-effect will also be executed twice.

```
// inside for await
const whenOnACertainState = state.as(CertainState)
if (whenOnACertainState) {
  await someSideEffect();  // this side effect will be executed 1 + the number of times the command call directly below fails
  await whenOnACertainState?.commands.someCommand();  
}
```

So you're right on doubting this one.

File: machine-runner/src/runner/runner.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (rkuhn) [reply machine-runner/src/runner/runner.ts:340]: No, when I said “state” I meant all of it, not just the state name. This would fix it, right?

File: machine-runner/src/runner/runner-utils.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (rkuhn) [reply machine-runner/src/runner/runner-utils.ts:40]: Wouldn’t it be confusing to have an observable that yields different results that the `for await` loop? Users can still use the audit events to get these triggers if needed.

File: machine-runner/src/runner/runner.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (Kelerchian) [reply machine-runner/src/runner/runner.ts:340]: Ah, do you mean reference comparison, as in `{} !== {}`?

[the remark this answers is not in the document]
Comment (rkuhn) [reply machine-runner/src/runner/runner.ts:340]: No, I mean value comparison, like `expect(x).toEqual(y)`. The same state can be recomputed due to time travel that has no effect, leading to new object references; emitting a “new” value would be confusing in this case.

File: machine-runner/src/runner/runner-utils.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (Kelerchian) [reply machine-runner/src/runner/runner-utils.ts:40]: It would be confusing if the user know beforehand that the `for await` loop trigger is the same as `on('change')`'s one.
Since the application code for the `for await` loop scenario and the `on('change')` scenario look totally different, it won't be confusing.

I think it is the other way around that if we want the user to subscribe to several event emitters, it would be a bit rigorous, especially since these separate event 'caughtUpAfterChange',  `command-rejection`, `command-issuance`, all of those are the internal mechanisms which are not visible to the eyes of the user.

[the remark this answers is not in the document]
Comment (rkuhn) [reply machine-runner/src/runner/runner-utils.ts:40]: Hmm. These emitter event names are part of our API, so we need to make sure that they convey their meaning as accurately as possible. `change` is quite unspecific, and it currently also doesn’t emit on every conceivable change (i.e. it doesn’t emit when the `state.commands` turns `undefined` due to a command having been invoked.

`caughtUpAndChanged` include implementation details that users should probably not need to be aware of.

---

In summary:
- we need to determine which events are actually needed right now — we shouldn’t add events just because someone might use them in the future
- each event we do emit needs to have a name that gives the user the right intuition of when it is emitted

File: machine-runner/src/runner/runner.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (Kelerchian) [reply machine-runner/src/runner/runner.ts:340]: To use a mechanism similar to `toEqual` would need an assumption on the data inside `payload`, which should be on the scope of the user. Other than that, jest's deep equality mechanism is custom-made, which means we need to make one too.

I see now that time-travel is one concern here. I think it is more accurate to track the last event that triggers the transition, before `caughtUp`. If before and after time travel, the last events of both have different eventId, trigger `state change`

File: machine-runner/src/runner/runner-utils.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (Kelerchian) [reply machine-runner/src/runner/runner-utils.ts:40]: `change` is more of a culture in the frontend world in a sense that `onChange` hook will be called if there is any meaningful change in the component that triggers the hook.

And we do have a `change` emission on `runner.ts` line 287. That is fired right after a `command` is called. This might be what makes it different between awaiting and not awaiting commands

[the remark this answers is not in the document]
Comment (Kelerchian) [reply machine-runner/src/runner/runner-utils.ts:40]: > each event we do emit needs to have a name that gives the user the right intuition of when it is emitted

What about `next`? It is unintuitive but we can make a promise that the AsyncIterable will be triggered on `next`, and then present the individual cases of when "next" is fired in the user manual.

File: machine-runner/src/runner/runner.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (rkuhn) [reply machine-runner/src/runner/runner.ts:340]: Time travel can also result in different events being processed before the final one, so I think we can only judge based on the state and its payload. Isn’t it established practice that deepEquals extends the value equality from primitive types to arrays and objects, recursively? That should be enough, I think.

[the remark this answers is not in the document]
Comment (rkuhn) [reply machine-runner/src/runner/runner.ts:340]: Since we’ll implement our `deepEquals` we can fix the behaviour — e.g. we could also compare Map and Set using value equality. We only need to document it.

File: machine-runner/src/runner/runner-utils.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (Kelerchian) [reply machine-runner/src/runner/runner-utils.ts:40]: > does putting it on the emitter method overload work

I don't quite understand what is emitter method overload. Could you elaborate?

File: machine-runner/src/runner/runner.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (Kelerchian) [reply machine-runner/src/runner/runner.ts:340]: One more alternative: How about:
- enabling user to write their own equality?
- use deepEqual as default mechanism?

[the remark this answers is not in the document]
Comment (rkuhn) [reply machine-runner/src/runner/runner.ts:340]: We should start without configuration options and only add them when we see that they’re needed.

[the remark this answers is not in the document]
Comment (rkuhn) [reply machine-runner/src/runner/runner.ts:340]: a cursory glance suggests that that implementation should treat Date specifically (but I haven’t reviewed whether correctly) — so we should use that impl?

[the remark this answers is not in the document]
Comment (Kelerchian) [reply machine-runner/src/runner/runner.ts:340]: Not sure. It uses `require` which might be harmful to pure es6 code, but I'll try running dev-example for it

[the remark this answers is not in the document]
Comment (Kelerchian) [reply machine-runner/src/runner/runner.ts:340]: @rkuhn there's a complication with equality check. 
It is possible that the application code, on a `reactIntoSelf`, simply returns a mutated old state. e.g. 

```
// type { someMap: Map() }

// reaction handler
(ctx) => {
  ctx.self.someMap.delete(someKey)
  return ctx.self
}
```

If we compare the equality this way:
```
const prev = internals.current
// mutation happens here
const next = internals.current
const changed = deepEqual(prev,next)
```

^ This will return true

If we do it this way:
```
const prev = deepCopy(internals.current)
// mutation happens here
const next = internals.current
const changed = deepEqual(prev,next)
```

This is expensive and prone to breakage when there is a new data type. There are data types already non-copyable such as WeakMap, WeakSet. Good `deepCopy` implementation is hard to come by.

File: machine-runner/src/runner/runner-utils.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (rkuhn) [reply machine-runner/src/runner/runner-utils.ts:40]: The `.on('x', ...)` methods all have the same name but different signatures, distinguished by their first argument. I think I’ve seen them separately documented.

File: machine-runner/src/runner/runner.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (rkuhn) [reply machine-runner/src/runner/runner.ts:340]: Let’s not overcomplicate things: we choose one deepEquals that we find reasonable and document that we use exactly that. API users will have to create their state payloads so that they get new states as desired.

[the remark this answers is not in the document]
Comment (Kelerchian) [reply machine-runner/src/runner/runner.ts:340]: Wait, I must assert that the complication happens due to the equality check (either deep or not):
- equality checks need 2 values in the first place
- state payload can be non-object

Therefore, either 1.) a copy needs to happen before the comparison, or 2.) the user is told to create their own new value.

> API users will have to create their state payloads so that they get new states as desired.

By this, do you mean option 2?

[the remark this answers is not in the document]
Comment (rkuhn) [reply machine-runner/src/runner/runner.ts:340]: Ah, I misunderstood which point you were talking about: yes, the copy before running the computation is needed, of course. We’ll also need to document that so people know what to expect. I think it is an anti-pattern to have large state payloads, because that implies having large events as well.

File: machine-runner/src/runner/runner-utils.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (Kelerchian) [reply machine-runner/src/runner/runner-utils.ts]: > does putting it on the emitter method overload work?
If this means, `next` and `change` behave differently, then it does.

File: machine-runner/src/runner/runner.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (Kelerchian) [reply machine-runner/src/runner/runner.ts]: Changed.
Copy and comparison now only happens on `caughtUp`, not at every computation, which is nice!

File: machine-runner/tests/esm/runner.test.ts [hunk 3]
```diff
@@ -353,7 +352,83 @@ describe('machine as async generator', () => {
...
+    await r1.feed([], true)
+    expect(promise1Watcher.isFinished()).toBe(true) // true because this is the first caughtUp
+
+    const promise2 = machine.next()
+    const promise2Watcher = watchPromise(promise2)
+    expect(promise2Watcher.isFinished()).toBe(false)
+    await r1.feed([], true)
+    expect(promise2Watcher.isFinished()).toBe(false) // false because there is no change
+
+    await issueCommand(machine.get())
```
Comment (rkuhn) [inline machine-runner/tests/esm/runner.test.ts:384]: this code relies on event loop details — perhaps we can make it more obvious using the following before asserting `.isFinished()` values:

```typescript
await new Promise((res) => setTimeout(res, 1))
```

File: machine-runner/tests/esm/runner.test.ts [hunk 4]
```diff
@@ -353,7 +352,83 @@ describe('machine as async generator', () => {
...
+    expect(promise2Watcher.isFinished()).toBe(true) // true because "next" event
+
+    const promise3 = machine.next()
+    const promise3Watcher = watchPromise(promise3)
+
+    await r1.timeTravel()
+    expect(promise3Watcher.isFinished()).toBe(false)
+    await r1.feed([], true)
+    expect(promise3Watcher.isFinished()).toBe(true) // caught up after timeTravel will trigger "next" event
```
Comment (rkuhn) [inline machine-runner/tests/esm/runner.test.ts:393]: I was confused at first: please add that the state is now different, hence we’re yielding it

File: machine-runner/tests/esm/runner.test.ts [hunk 5]
```diff
@@ -353,7 +352,83 @@ describe('machine as async generator', () => {
...
+    expect(promiseWatcher.isFinished()).toBe(false)
+
+    await r1.feed([ToggleOn.make({})], false)
+    // False because there is no 'caughtUp' event
+    expect(promiseWatcher.isFinished()).toBe(false)
+
+    await r1.feed([], true)
+    // True because a change happens between previous and current 'caughtUp'
+    expect(promiseWatcher.isFinished()).toBe(true)
+  })
```
Comment (rkuhn) [inline machine-runner/tests/esm/runner.test.ts:429]: :+1:

File: machine-runner/README.md [hunk 6]
```diff
@@ -214,6 +214,13 @@ If the workflow still is in the `Auction` state, we compute the best robot bid (
 The third feature becomes relevant once the auction has ended: we check if our robot is indeed the winner and record that in a variable `IamWinner`, i.e. in the current application in-memory state.
 Then we can use this information in all following states as well.
 
+### Change detection on for-await loop
+
+When using for-await loop with the machine runner, the loop iterates only if the following criteria matches:
```
Comment (rkuhn) [inline machine-runner/README.md:219]:
```suggestion
When using a for-await loop with the machine runner, the loop iterates only if all of the following criteria are met:
```

**Review by rkuhn — APPROVED** [review]
Very good! Just a few tiny nitpicks.

File: machine-runner/tests/esm/runner.test.ts (hunk 4 above)
Comment (Kelerchian) [reply machine-runner/tests/esm/runner.test.ts:393]: Ah, I just realized that it wasn't that obvious. Adding comments

File: machine-runner/tests/esm/runner.test.ts (hunk 3 above)
Comment (Kelerchian) [reply machine-runner/tests/esm/runner.test.ts:384]: Thanks for the catch
